翻譯 RFC 2342 const control flow

CYBAI opened the pull request:
Fix #1 

---

有語意不順或是覺得怪怪的地方都歡迎留言討論！ 🙇 

---

[Rendered](https://github.com/CYBAI/rfcs-tw/blob/2342-zh-tw/text/2342-const-control-flow.md)

CYBAI committed:
Translate RFC 2342

wusyong commented on text/2342-const-control-flow.md:
```diff
@@ -3,23 +3,24 @@
 
-Conditions in constants are important for making functions like `NonZero::new`
-const fn and interpreting assertions.
+在常數宣告中使用條件式對於建立像是 `NonZero::new` 的 `const fn` 及 直譯判定 (interpreting assertions) 來說很重要。
```
括號推薦使用全形（），這個才是正體中文的標點符號。

CYBAI replied:
感謝！忘記這件事了 😂

wusyong commented on text/2342-const-control-flow.md:
```diff
@@ -3,23 +3,24 @@
+在常數宣告中使用條件式對於建立像是 `NonZero::new` 的 `const fn` 及 直譯判定 (interpreting assertions) 來說很重要。
 
-# Guide-level explanation
+# Guide-level 解說
```
```suggestion
# 教學式解說
```

wusyong commented on text/2342-const-control-flow.md:
````diff
@@ -82,33 +78,25 @@ const FOO: SomeType = if X > Y {
 ```
 
-# Reference-level explanation
+# Reference-level 解說
````
```suggestion
# 技術文件式解說
```
這個和上面一樣只是建議而已，讓讀者可以直白地知道這段落主要的目的。

CYBAI replied:
感謝！其實我是想不到好的翻譯詞所以暫時先照原文 😂 不過你的兩個提議都不錯！

wusyong commented on text/2342-const-control-flow.md:
```diff
@@ -135,9 +123,7 @@ const fn foo(x: u32, y: u32) -> u32 {
-branch is evaluated.
+因為 const fn 的 `x` 和 `y` 參數是未知的，他們無法被做常數求值 (const evaluate)。當提供此 const fn 參數並求值時，只有相對應的分支會被求值。
 
 # Unresolved questions
 [unresolved]: #unresolved-questions
```
```suggestion
# 未解決問題
[unresolved]: #未解決問題
```

CYBAI replied:
> `[unresolved]: #未解決問題`

這個應該是 URL 的 hash 所以我就不改了 XD

wusyong reviewed (commented):
LGTM with a few suggestions. If you think they are not necessary  I can still approve and merge!

weihanglo commented on text/2342-const-control-flow.md:
```diff
@@ -63,12 +62,9 @@ const FOO: SomeType = if X > Y {
-`A` and `B` are evaluated before `FOO`, since constants are by definition
-constant, so their order of evaluation should not matter. This assumption breaks
-in the presence of errors, because errors are side effects, and thus not pure.
+`A` 和 `B` 會在 `FOO` 之前被求值，因為常數是根據定義而來的常數，所以求值的順序並不應該影響。這項假設在有錯誤的情況下並不成立，因為錯誤屬於副作用，因次不純 (pure)。
```
```suggestion
`A` 和 `B` 會比 `FOO` 先被求值，因為常數是根據定義而來的常數，所以求值的順序並不應該影響。這項假設在有錯誤的情況下並不成立，因為錯誤屬於副作用，因此不純 (pure)。
```

CYBAI replied:
感謝！

另外想問

> since constants are by definition constant

這句翻成「因為常數是根據定義而來的常數」總覺得哪裡怪怪的... 會嗎 >_<?

weihanglo replied:
我再抓了一次意思，比較像是「因為常數在定義上就是...常數，所以不應被求值順序影響」

CYBAI replied:
後面翻「所以不應被求值順序影響」感覺更順了 👍

> 因為常數在定義上就是...常數

啊！原來！！感謝 XD

CYBAI replied:
Fixed in 5ce4e49

weihanglo commented on text/2342-const-control-flow.md:
```diff
@@ -135,9 +123,7 @@ const fn foo(x: u32, y: u32) -> u32 {
-Since the const fn's `x` and `y` arguments are unknown, they cannot be const
-evaluated. When the const fn is evaluated with given arguments, only the taken
-branch is evaluated.
+因為 const fn 的 `x` 和 `y` 參數是未知的，他們無法被做常數求值 (const evaluate)。當提供此 const fn 參數並求值時，只有相對應的分支會被求值。
```
NIT：去掉「是 XX 的」比較沒那麼翻譯腔

```suggestion
 const fn 的 `x` 和 `y` 參數未知，無法做常數求值 (const evaluate)。當提供此 const fn 參數並求值時，只會對相應的分支求值。
```

CYBAI replied:
同意！「他們」也翻出來有點太照原文了 😂 感謝！

weihanglo commented on text/2342-const-control-flow.md:
````diff
@@ -120,7 +108,7 @@ const AB: u32 = if X > Y {
 ```
 
-where either `X - Y` or `Y - X` would emit an error, add an intermediate const fn
+`X - Y` 或是 `Y - X` 其中一方有可能會報錯，這時我們必須加入中介 const fn
````
NIT：語句上不加我們也通順

```suggestion
`X - Y` 或是 `Y - X` 其中一方有可能會報錯，這時必須加入中介 const fn
```

weihanglo commented on text/2342-const-control-flow.md:
```diff
@@ -82,33 +78,25 @@ const FOO: SomeType = if X > Y {
 [alternatives]: #alternatives
 
-## Require intermediate const fns to break the eager const evaluation
+## 利用中介 const fns 來破壞進擊的常數求值 (const evaluation)
```
eager evaluation 感覺很像 greedy 的感覺，我會想翻譯成
- 「貪婪的常數求值」或
- 「立即常數求值」

然後括號裡面應該放 _eager const evaluation_ 感覺上

CYBAI replied:
「立即常數求值」聽起來蠻符合他的意思 +1

weihanglo commented on text/2342-const-control-flow.md:
```diff
@@ -82,33 +78,25 @@ const FOO: SomeType = if X > Y {
 
-This makes it easier to fail compilation on random "constant" values like
-`size_of::<T>()` or other platform specific constants.
+這項功能會很容易的在任意「常數」值（例如 `size_of::<T>()` 或是特定的平台常數）造成編譯失敗。
```
動詞提到前面感覺更直接告知回「造成」這件事

```suggestion
這項功能容易造成任意「常數」值（如 `size_of::<T>()` 或特定平台常數）編譯失敗。
```

CYBAI replied:
天... 現在再讀一次我翻譯的第一版也太難懂... orz 感謝！！！

weihanglo reviewed (commented):
Basically looks great! 

Some sentences are not in idiomatic Chinese such like using too many passive voice. Still can move on without concerning my suggestions. 👍🏾 👍🏾 👍🏾 👍🏾

CYBAI commented:
Thanks for the review from both of you 🙇 I just addressed all the comments; if it looks fine, I will squash the commits (or maybe use `squash and merge`?)!

CYBAI commented:
I added the `rendered` link in PR description 🙇‍♂️

weihanglo commented on text/2342-const-control-flow.md:
```diff
@@ -82,33 +78,24 @@ const FOO: SomeType = if X > Y {
 [alternatives]: #alternatives
 
-## Require intermediate const fns to break the eager const evaluation
+## 利用中介 const fns 來破壞立即常數求值 (eager const evaluation)
```
```suggestion
## 利用中介 const fns 來破壞立即常數求值（eager const evaluation）
```

weihanglo commented on text/2342-const-control-flow.md:
```diff
@@ -135,9 +122,7 @@ const fn foo(x: u32, y: u32) -> u32 {
 
-# Unresolved questions
+# 未解決問題
 [unresolved]: #unresolved-questions
```
這個是 mdbook 產生 intra-doc 的 reference link 的 anchor，不改的話 `[a-ref-link][unresolved]` 並沒辦法找到對應的「未解決問題」這一節，不過其實整篇 RFC 沒用到，說實話完全不影響功能

```suggestion
[unresolved]: #未解決問題 
```

CYBAI replied:
Heh, 原來！第一次知道 @@! 其他的部分也一起改一改好了，因為我想未來其他翻譯的 RFC 也一起統一格式（都有翻譯）比較好 🤔? 感謝！

weihanglo reviewed (commented):
Commented. The rest LGTM.

CYBAI commented:
hmm, I'm not so used to no CI tests 😂 I dunno what we should test though...

Anyway, I will merge directly this time and maybe we can discuss further if anyone has any ideas.

Also, thanks for the reviews! @weihanglo @wusyong 🙏
